### ISSUE TITLE
ttnn.sort: CrossCoreDataExchange unstable sorts return invalid indices on ties (duplicates in tie groups; out-of-range indices >= logical W at non-pow2 widths)

### ISSUE THREAD

**Issue by nkapreTT** [tenstorrent/tt-metal#54043]
### Summary

Unstable (`stable=False`) `ttnn.sort` calls routed to the **CrossCoreDataExchange** program factory return **invalid index tensors** whenever a value tie group spans a cross-core tile pair. Two manifestations, same root cause:

1. **Duplicate indices inside tie groups** at CrossCore's native widths (padded W = 4096 / 8192): the index tensor is not a permutation of `[0, W)`. Values are sorted exactly and `torch.gather(input, -1, indices)` still reproduces the sorted values (both duplicated entries carry the tied value), so value-only and gather-parity tests pass while the indices are silently wrong.
2. **Out-of-range indices at non-power-of-2 logical widths**: indices ≥ logical W are returned (they point into the ±inf pad lanes added by the composite padding), and ascending-value gather verification fails. **This is an out-of-bounds hazard for downstream `gather` / `scatter` / embedding-lookup consumers** — the returned index is outside the logical tensor extent.

Observed on Blackhole p150a, current `main` (pre-dates any in-flight sort work — reproduced on `main` as well as feature branches). Affects **both TILE and ROW_MAJOR** layouts, H from 32 to 320, both `descending` values. The stable path (`stable=True`, index-aware comparator `topk_cmp_swap_stable_directional`) is NOT affected — it survived a 4,300+-cell adversarial sweep with exact `torch.sort(stable=True)` parity.

### Minimal repro A — duplicate indices in tie groups (W=4096, also fires at 8192)

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

x = torch.full((32, 4096), 1.0).bfloat16()  # one giant tie group
tx = ttnn.from_torch(x, ttnn.bfloat16, layout=ttnn.Layout.TILE, device=device)
values, indices = ttnn.sort(tx, dim=-1, stable=False)

idx = ttnn.to_torch(indices).to(torch.int64)
perm = torch.arange(4096).expand(32, -1)
rows_ok = (torch.sort(idx, dim=-1).values == perm).all(dim=-1)
print(rows_ok.sum().item(), "/ 32 rows are a permutation")  # 0 / 32 — every row has duplicates
# values are exact and torch.gather(x, -1, idx) == sorted values, so the
# corruption is invisible to value/gather checks.
```

Same result on `ttnn.Layout.ROW_MAJOR` and with partial tie groups (e.g. 5-level tie-heavy inputs): duplicates appear inside each tie group.

### Minimal repro B — indices ≥ logical W at non-pow2 width (OOB hazard)

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

W = 3000  # pads to 4096 with ±inf sentinels at the composite layer
x = torch.full((32, W), -2.0)
x[:, 0] = float("inf")
x[:, W - 1] = float("-inf")
tx = ttnn.from_torch(x.bfloat16(), ttnn.bfloat16, layout=ttnn.Layout.TILE, device=device)
values, indices = ttnn.sort(tx, dim=-1, stable=False)

idx = ttnn.to_torch(indices).to(torch.int64)
print(int(idx.max()), ">= W =", W)  # index 3011 >= 3000, observed in EVERY row
# torch.gather on the logical region also mismatches the returned sorted values
# (a pad-lane index was returned in place of a real-element index).
```

A downstream consumer doing `torch.gather(some_tensor, -1, idx)` (or an on-device gather/scatter) reads out of the logical bounds.

### Mechanism (from code reading of `sort_cross_core_data_exchange.cpp` + `ckernel_sfpu_topk.h`, confirmed on silicon)

- The unstable exchange path uses raw `TTI_SFPSWAP` with hardware index tracking (SFPCONFIG bit 2). On an exact value tie SFPSWAP does **not** swap, so each element's index stays with its DEST slot — **tie resolution is positional**.
- In the cross-core exchange branch, **both peer cores of a spanning tile pair independently run `topk_merge` with swapped operand order**: each loads its LOCAL tile into DEST slot 0 (`input_dest_start`) and the PEER tile into slot 1. The positions of the same two elements are therefore opposite on the two cores.
- Both cores compute the same block direction (`i` and `j` differ only below bit `stage`, so `(i >> stage) & 1` is identical) but keep **opposite halves** (`select_lower = dir ^ (i < j)` flips between the peers).
- Worked example (dir=0, tied lane with `(x, idx_a)` on tile i and `(x, idx_b)` on tile j): core-i's merge leaves `DEST[0]=(x,a), DEST[1]=(x,b)`, `select_lower=1` → keeps `idx_a`; core-j's merge leaves `DEST[0]=(x,b), DEST[1]=(x,a)`, `select_lower=0` → keeps `DEST[1]` = `idx_a`. **Both emit `idx_a`; `idx_b` vanishes.**
- The mechanism is width-independent — it fires whenever CrossCore runs multi-core and a tie spans a pair (`get_number_of_tiles_per_core(USE_AS_MANY_CORES)` = 2 tiles/core at these widths → spanning pairs at every `sub_dist >= 2`). W=4096/8192 are simply the widths the CrossCore factory natively owns in `select_program_factory`.
- Repro B is the same defect escalated by the composite's ±inf pad lanes: pad sentinels tie with each other (and with real ±inf data), and a pad-lane index wins a tie against a real-element index.
- The purely local paths (SingleRowSingleCore; CrossCore's same-core pair branch) move value+index atomically per SFPSWAP and always emit a valid permutation — only the cross-core spanning-pair exchange is affected.

### Expected behavior

`torch.sort(..., stable=False)` makes no promise about *which* index inside a tie group appears at a given output position, but the index tensor must still be a valid permutation of `[0, W)` per row, and every index must be `< W` (logical). Both properties are violated here.

Filing for visibility/tracking — a fix would need either an index-aware compare in the unstable exchange (as the stable comparator does) or a canonical single-peer operand order for spanning pairs.
